### PR TITLE
interop updates

### PR DESCRIPTION
**Pull request by arnetheduck**
* add interop launcher scripts
* stick validator_keygen into beacon_node
* fix lmd ghost slot number on missing block
* use mocked eth1data when producing blocks
* use bls public key method for withdrawal credentials
* fix deposit domain
* prefer lowercase for a bunch of toHex
* build simulation binary in data folder to avoid data types confusion

### CODE REVIEW

**Comment by zah** [comment]
I've already implemented the mocked start in an earlier PR here:
https://github.com/status-im/nim-beacon-chain/pull/357

I see some discrepancies with the spec in the code here and the command-line handling in my PR is done is a slightly better way, so I'll try to finish it today and merge it, so this PR can be rebased on top of it.

**Comment by arnetheduck** [comment]
looking at it #357, I suspect the actual discrepancies are fewer in this branch and this branch has better long-term code for them (ie withdrawal credentials, interop eth1 data, deposit construction with the right signature domain etc). the launcher scripts here probably are nicer too, though we can update them to whichever command line interface we end up using.

re command line, I found it quite bothersome that two subcommands can't have the same parameter (`validatorCount` for example) - any way around that?

**Comment by zah** [comment]
> re command line, I found it quite bothersome that two subcommands can't have the same parameter (validatorCount for example) - any way around that?

That's a limitation of Nim's case objects. You can work-around it by using a different name for the field itself (e.g. `quickStartValidatorCount`), but with the same `longform` and `shortform` names for the option (these are specified with pragmas).

File: beacon_chain/beacon_node.nim [hunk 1]
```diff
@@ -793,13 +791,17 @@ when isMainModule:
...
-      uint64(times.toUnix(times.getTime()) + config.genesisOffset),
-      Eth1Data(), {skipValidation})
+
+    var
+      startTime = uint64(times.toUnix(times.getTime()) + config.genesisOffset)
+      initialState = initialize_beacon_state_from_eth1(
+        eth1BlockHash, startTime, deposits, {skipValidation})
+
+    # https://github.com/ethereum/eth2.0-pm/tree/6e41fcf383ebeb5125938850d8e9b4e9888389b4/interop/mocked_start#create-genesis-state
+    initialState.genesis_time = startTime
```
Comment (zah) [inline beacon_chain/beacon_node.nim]: This assignment is not necessary. `initialize_beacon_state_from_eth1` already sets the `genesis_time` to `startTime`:

https://github.com/status-im/nim-beacon-chain/blob/master/beacon_chain/spec/beaconstate.nim#L227

Comment (arnetheduck) [reply beacon_chain/beacon_node.nim]: not any more: https://github.com/status-im/nim-beacon-chain/pull/371/files#diff-8491da098fcb79b0774953064b352de1R228

File: beacon_chain/conf.nim [hunk 2]
```diff
@@ -131,6 +132,16 @@ type
...
 
+    of makeDeposits:
+      totalDeposits* {.
+        desc: "Total number of deposits and keys to generate".}: int
+
+      depositDir* {.
+        desc: "Folder to write deposits to", defaultValue: "validators".}: string
+
+      randomKeys* {.
+        desc: "Use random keys (instead of interop keys)", defaultValue: false.}: bool
```
Comment (zah) [inline beacon_chain/conf.nim]: At some point, the default here should be switched to `true`. We can leave a TODO comment about it.

File: interop/README.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+In general, follow the build instructions of `nim-beacon-chain` as documented in the main repo - make sure to set up your build environment with all necessary system libraries as documented there:
+
+```bash
+# Clone repo
+git clone https://github.com/status-im/nim-beacon-chain.git
+
+cd nim-beacon-chain
+
+make # prepare build system (cloning the correct submodules)
+make update deps # build dependencies
```
Comment (zah) [inline interop/README.md]: I think it's just `make update` now.

Comment (stefantalpalaru) [reply interop/README.md]: Yep.

Comment (arnetheduck) [reply interop/README.md]: oh. I thought I saw a `deps` target in the makefile - so I guess that's called automatically..?

Comment (stefantalpalaru) [reply interop/README.md]: It would be, if start.sh would use the Makefile to build beacon_node.

File: beacon_chain/conf.nim (hunk 2 above)
Comment (arnetheduck) [reply beacon_chain/conf.nim]: thought about this, but it's slightly weird - you would never generate random keys en masse like that in a real network, so what remains is testing - and do we really want random keys then? or should we be deterministic with some parameter maybe that changes the seed?

File: beacon_chain/fork_choice.nim [hunk 4]
```diff
@@ -53,9 +50,9 @@ proc lmdGhost*(
 
     head = head.children[0]
     var
-      headCount = get_vote_count(head)
+      headCount = get_latest_attesting_balance(head)
 
     for i in 1..<head.children.len:
-      if (let hc  = get_vote_count(head.children[i]); hc > headCount):
+      if (let hc  = get_latest_attesting_balance(head.children[i]); hc > headCount):
```
Comment (mratsim) [inline beacon_chain/fork_choice.nim]: does that supercede https://github.com/status-im/nim-beacon-chain/pulls?

**Pull request merged**